### PR TITLE
[GitHub][workflows] Add buildbot information comment to first merged PR from a new contributor

### PR DESCRIPTION
**Pull request by DavidSpickett**
This change adds a comment to the first PR from a new contributor that is merged, which tells them what to expect post merge from the build bots.

How they will be notified, where to ask questions, that you're more likely to be reverted than in other projects, etc. The information overlaps with, and links to https://llvm.org/docs/MyFirstTypoFix.html#myfirsttypofix-issues-after-landing-your-pr. So that users who simply read the email are still aware, and know where to follow up if they do get reports.

To do this, I have added a hidden HTML comment to the new contributor greeting comment. This workflow will look for that to tell if the author of the PR was a new contributor at the time they opened the merge. It has to be done this way because as soon as the PR is merged, they are by GitHub's definition no longer a new contributor and I suspect that their author association will be "contributor" instead.

I cannot 100% confirm that without a whole lot of effort and probably breaking GitHub's terms of service, but it's fairly cheap to work around anyway. It seems rare / almost impossible to reopen a PR in llvm at least, but in case it does happen the buildbot info comment has its own hidden HTML comment. If we find this we will not post another copy of the same information.

### CODE REVIEW

**Comment by DavidSpickett** [comment]
Important to note that this comment will notify all users, there's no way to avoid that. I hope that by `@` ing the author, it may be more friendly to email filters.

**Comment by DavidSpickett** [comment]
https://discourse.llvm.org/t/rfc-buildbot-information-comment-on-new-contributors-first-prs/76303 to get more diverse feedback.

File: .github/workflows/merged-prs.yml
Comment (tstellar) [inline .github/workflows/merged-prs.yml:26]: Can you change this to use a sparse checkout like we do in https://github.com/llvm/llvm-project/blob/eca25bbdbb4f4681cd82de80e92b5f1315f0996a/.github/workflows/build-ci-container.yml#L39

Comment (DavidSpickett) [reply .github/workflows/merged-prs.yml]: Done.

Would it be useful if I updated the other workflows that use curl also?

Comment (DavidSpickett) [reply .github/workflows/merged-prs.yml]: (in new PRs ofc)

Comment (tstellar) [reply .github/workflows/merged-prs.yml]: Yes, that would be great.

File: .github/workflows/merged-prs.yml
Comment (tstellar) [inline .github/workflows/merged-prs.yml:26]: You also need to specify a branch here so that it checks the code out from the main branch and not the PR.

File: .github/workflows/merged-prs.yml
Comment (boomanaiden154) [inline .github/workflows/merged-prs.yml:32]: Is there a reason this isn't `chmod +x` in the tree? We could also just run it with a `python3` prefix instead of making it executable.

File: llvm/utils/git/github-automation.py
Comment (boomanaiden154) [inline llvm/utils/git/github-automation.py:283]: Maybe add a link to https://lab.llvm.org here or somewhere below and indicate how to check the builds for a specific change?

File: llvm/utils/git/github-automation.py
Comment (boomanaiden154) [inline llvm/utils/git/github-automation.py:296]: I would add a note here that the policy is to open a new PR relanding the changes.

**Review by boomanaiden154 — COMMENTED** [review]
A couple minor comments. Thanks for working on this!

File: .github/workflows/merged-prs.yml
Comment (DavidSpickett) [reply .github/workflows/merged-prs.yml]: I've used `python3`, if that's cool with you I'll do that for the other workflows in another PR.

File: llvm/utils/git/github-automation.py
Comment (DavidSpickett) [reply llvm/utils/git/github-automation.py]: There's no great way to do that unfortunately.

The console view (https://lab.llvm.org/buildbot/#/console) is the least bad, if your change has been merged recently. Recently meaning in the last hour, right now that's the oldest change I see. Might be even shorter of a window in peak times like the US working day.

Most new contributors won't have commit access also, so it's more likely that their change gets merged while they are busy with other things.  But it's the least bad option, and if we make them aware of it they can use it later when they are able to do their own merges.

So I've added a note about the console view to the text.

The other option is the changes page e.g. https://lab.llvm.org/buildbot/#/changes/120892 however that requires you to know the Buildbot change ID for the commit, I couldn't find a way to get that from a git hash. Also, that lists all builds it's been included in, including builds that were already red (for example https://lab.llvm.org/buildbot/#/builders/272/builds/5779). Which would be more confusing than useful, even if we had a way to get from git hash to buildbot change ID.

File: llvm/utils/git/github-automation.py
Comment (DavidSpickett) [reply llvm/utils/git/github-automation.py]: Added this.

File: llvm/utils/git/github-automation.py
Comment (boomanaiden154) [reply llvm/utils/git/github-automation.py]: I prefer the last changes page (https://lab.llvm.org/buildbot/#/changes) over the console, but it does have the same issue (limited to the last 50 changes). The current rendition for this LGTM though.

I think you can get the buildbot ID by checking the CI status on Github on the commit in `main` as the buildbot statuses get reported there and link to the individual builders. Not super trivial though and having directions on copying around IDs just doesn't seem helpful.

**Review by boomanaiden154 — APPROVED** [review]
LGTM

File: llvm/utils/git/github-automation.py
Comment (DavidSpickett) [reply llvm/utils/git/github-automation.py]: This runs on merged PRs, so it might be possible to go from git hash -> GitHub -> buildbot statuses -> change ID and make a customised link to go in the comment. I'll look into that for a future change.

**Comment by DavidSpickett** [comment]
Thanks for reviewing! I'm going to wait a few days for any feedback from the RFC before landing.

**Comment by DavidSpickett** [comment]
> Maybe that's getting a bit lengthy though...

Given we have so many "but this", "sometimes that" I'm wondering if it's not better for me to update one of the documentation pages, then link to that. Then we have more room to include them. Not sure which yet though.

**Comment by DavidSpickett** [comment]
> "If you are unsure how to fix a problem, you can send questions in a reply to the notification email". If your change is the only one causing a failure, then the only person to reply to is gkistanova (I think [llvm.buildmaster@lab.llvm.org](mailto:llvm.buildmaster@lab.llvm.org) just bounces). Otherwise, you're replying to other patch authors in the change list. Perhaps directing people to the PR or to discord would be better?

Yeah, not mentioning the email reply is probably better. The reviewers of the PR are more likely to have context on why it failed (I would hope), and they can sort it out from there.

**Comment by RalfJung** [comment]
> We do seem to have some buildbots that have spurious failures every now and again. I think it would be good to warn of that possibility. e.g. "On occasion buildbots will encounter a spurious failure (i.e. failing for a reason other than your change). 

So far, I think I got buildbot failures after 100% of my PRs that landed here (which is 5). My PRs were all changing the LangRef and nothing else, so none of them was a real failure caused by my PR.

So I wouldn't say this is a rare occurrence, it seems to be the common case. My conclusion is that I should just entirely ignore these emails... that's probably not right either though, but I am not sure how I could even figure out whether I need to take any action.

**Comment by DavidSpickett** [comment]
Thanks for your feedback, this is exactly what I was looking for!

> So I wouldn't say this is a rare occurrence, it seems to be the common case.

Yeah my bias comes in here seeing builds go by all the time that do work properly, I forget the important bit is the builds the specific committer sees. So your take away isn't wrong at all. Unfortunate, but not a failure on the part of committers, instead a sign that the system should be improved.

> So I wouldn't say this is a rare occurrence, it seems to be the common case. My conclusion is that I should just entirely ignore these emails... that's probably not right either though, but I am not sure how I could even figure out whether I need to take any action.

We're asking a lot of folks to understand a whole other system, and explaining how to diagnose problems generally is too much to include here and something I wouldn't want folks to think they've got to spend too much time on.

Your feedback certainly backs up Alex's suggestions so I will add those now.

**Comment by DavidSpickett** [comment]
I've updated the text with the feedback so far.

I think it could go on https://llvm.org/docs/GitHub.html instead as "what happens after merge" so going to look at that next.

**Comment by DavidSpickett** [comment]
@asb I've updated the text to link to https://llvm.org/docs/MyFirstTypoFix.html#myfirsttypofix-issues-after-landing-your-pr, but there's still a considerable overlap as you suggested.

My intent is that the text in the comment warns you about the initial messages you might get, then tells you, if you need to follow up on them, here's the link to a description of all that.

**Comment by asb** [comment]
New text LGTM, but it looks like some of the newly added lines aren't line-wrapped while the existing lines are.

**Comment by DavidSpickett** [comment]
Yeah so that's an annoying thing I've had to work around. Unlike RST the lines don't flow automatically regardless of the newlines. So I've tried to lay it out so that the paragraphs aren't too weird looking, but the text is still readable in an editor.

Also when you have a `[link](giant URL...............................goes here)` that renders into a single word `link`. So also tried to write
it so we'd end up with a full line there. Instead of:
```
Some text with a <link>.
Next sentence ends up here which looks strange given it's supposed to be part of the same paragraph
```

**Comment by DavidSpickett** [comment]
Added a comment to explain that oddity.

**Review by asb — APPROVED** [review]
LGTM. I'd be OK with just having the comment in unwrapped form so that it's stored in the preferred form to be submitted as a GitHub comment, but the trade-off you make seems reasonable too.

**Pull request merged**
